specs/agents/tracing-spans-destination.md

```diff
@@ -0,0 +1,101 @@
+If the span represents a call to an in-memory database, the information SHOULD still be set.
+
+Agents SHOULD have a generic component used in all tests that validates that the destination information is present for exit spans.
+Rather than opting into the validation, the testing should provide an opt-out if,
+for whatever reason, the destination information can't or shouldn't be collected for a particular exit span.
+
+#### `context.destination.service.name`
+
+ES field: `span.destination.service.name`
+
+The identifier for the destination service.
+
+**Usage**
+
+Currently, this field is not used anywhere within the product.
+The original intent was to use it as a display name of a service in the service map.
+
+**Value**
+
+For HTTP, use scheme, host, and non-default port (e.g. `http://elastic.co`, `http://apm.example.com:8200`).
+For anything else, use `span.subtype` (e.g. `postgresql`, `elasticsearch`).
+If cluster information is available, it should be appended `${span.subtype}/${cluster}`.
+However, individual sub-resources of a service, such as the name of a message queue, should not be added.
+
+#### `context.destination.service.resource`
+
+ES field: `span.destination.service.resource`
+
+Identifies unique destinations for each service.
+
+**Usage**
+
+Each unique resource will result in a node on the service map.
+Also, APM Server will roll up metrics based on the resource.
+These metrics are currently used for the [dependencies table](https://www.elastic.co/guide/en/kibana/current/service-overview.html#service-span-duration)
+on the service overview page.
+There are plans to use the service desination metrics in the service map, too.
+
+The metrics are calculated based on the (head-based) sampled span documents that are sent to APM Server.
+That's why agents have to send the [`sample_rate`](tracing-sampling.md#effect-on-metrics)
+attribute for transactions and spans:
+It is used by APM Server to extrapolate the service destination metrics based on the (head-based) sampled spans.
+
+**Cardinality**
+
+To avoid a huge impact on storage requirements for metrics,
+and to not "spam" the service map with lots of fine-grained nodes,
+the cardinality has to be kept low.
+However, the cardinality should not be too low, either,
+so that different clusters, instances, and queues can be displayed separately in the service map.
+
+The cardinality should be the same or higher as `span.destination.service.name`.
+Higher, if there are individual sub-resources for a service, such as individual queues for a message broker.
+Same cardinality otherwise.
+
+**Value**
+
+Generally, the value would look something like `${span.subtype}/${cluster}`, or `${span.subtype}/${queue}`.
+For HTTP, this is the address, with the port (even when it's the default port), without any scheme.
+The specs for the specific technologies will have more information on how to construct the value for `context.destination.service.resource`.
```

Comment:
Does this include the path?

Reply:
Do you mean whether the `context.destination.service.resource` contains the URL path for HTTP spans? No, just host and port. See also the destination part of the HTTP spec.
Including the path would be problematic in terms of cardinality.

Reply:
Can it _never_ include the path, even if we e.g. know that we are talking to an API gateway that describes the service as the first part of the path?

Reply:
In those specific scenarios, it can make sense to include the path. But with generic instrumentation, it's hard to detect these specific cases.

Reply:
Agree that custom instrumentation would be needed. Does it make sense to mention that in the spec?

Reply:
This spec is mainly targeted towards the auto-instrumentations for agents. For that reason, I think it'd leave that out here. But it's a good point in favor of adding a public API for destination #424. I was thinking that we might not even need such n API if we can automatically derive these fields. But offering customization does make sense, to allow for the use case you mentioned. But I'd still expect most users to be fine with auto-derived destination fields.